### PR TITLE
JME related additions and bug fixes

### PR DESCRIPTION
**Pull request by rappoccio**
A. There are a few additions needed for JEC and JER development:
1. Out of time pileup before and after "this" bunch crossing. 
2. Generator weight
3. Generator pt-hat (when present)

B. There are also two bug fixes:
1. Energy fractions were wrong (using corrected energy in denominator, raw in numerator of fractions). 
2. Soft dropped mass did not have jet corrections applied. (WAITING FOR https://github.com/cms-sw/cmssw/pull/21901 BEFORE THIS IS ADDED)

For A1: I implemented a new producer for the NPU variables (NPUTablesProducer). 

For B2: I implemented an ease-of-use accessor in pat::Jet to get the groomed mass. (WAITING FOR https://github.com/cms-sw/cmssw/pull/21901 BEFORE THIS IS ADDED)

### CODE REVIEW

**Comment by arizzi** [comment]
changing the dataformat may make it harder to integrate

wasn't there already a producer for NPU? can you pleaae extend that?

**Comment by rappoccio** [comment]
This isn’t really a dataformat change, it just adds a method. 

I didn’t see an NPU producer... it was a flat table producer, unless I missed something???

**Comment by arizzi** [comment]
SimplePileupFlatTableProducer
see globals.py

File: DataFormats/PatCandidates/interface/Jet.h
Comment (gpetruc) [inline DataFormats/PatCandidates/interface/Jet.h]: please use `const std::string &` instead of `std::string`
(it should be changed also in the other methods, BTW)

File: PhysicsTools/NanoAOD/plugins/NPUTablesProducer.cc
Comment (gpetruc) [inline PhysicsTools/NanoAOD/plugins/NPUTablesProducer.cc]: I'd use int instead of uint8.
The disk space after compression should be the same, and we don't risk to run into problems if we generate high PU samples.
Also, since you are at it, can you change the nTrueInt to float, as per https://github.com/cms-nanoAOD/cmssw/issues/95

File: PhysicsTools/NanoAOD/python/jets_cff.py
Comment (gpetruc) [inline PhysicsTools/NanoAOD/python/jets_cff.py]: maybe it's safer by label, in case we change at some point the upstream configuration

File: PhysicsTools/NanoAOD/plugins/NPUTablesProducer.cc
Comment (gpetruc) [inline PhysicsTools/NanoAOD/plugins/NPUTablesProducer.cc]: Since you add this, can you remove the instantiation of 
SimplePileupFlatTableProducer from https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/plugins/SimpleFlatTableProducer.cc#L236 which will no longer be needed?

**Review by gpetruc — CHANGES_REQUESTED** [review]

**Comment by rappoccio** [comment]
OK, all implemented.

**Review by gpetruc — APPROVED** [review]

File: DataFormats/PatCandidates/interface/Jet.h
Comment (gpetruc) [inline DataFormats/PatCandidates/interface/Jet.h]: I didn't notice it in the previous iterations, but I suppose the `(0)` is meant to be `(index)`, right?

scram b code-checks may complain about using `x.size() > 0` instead of `!x.empty()` (here and in the next line), maybe you can change that as well?

**Comment by rappoccio** [comment]
Doh! You are correct, now this is fixed. Sorry about that!

**Comment by gpetruc** [comment]
retrying after patching the build script to avoid rebuilding fireworks (which seems to require libs not in the standard docker + cmvfs environment).

That said, given that the change in DataFormats/PatCandidates triggers the recompilation of lots of packages, it may be useful to place already the PRs to CMSSW and ask if it can be considered for the next CMSSW_9_4_X release.
If this branch is not yet merged anywhere else, then I would probably suggest to re-write the history and split this into:
 * one NanoAOD-only PR that does not touch DataFormats/PatCandidates, made to this repository, that we can merge it quickly
 * one with just the DataFormats/PatCandidates addition and the groomed mass change, to be made to the CMSSW repo (and that can we merge also here once it is accepted there)

**Comment by rappoccio** [comment]
OK DataFormats PRs are:
https://github.com/cms-sw/cmssw/pull/21901
https://github.com/cms-sw/cmssw/pull/21902

**Comment by rappoccio** [comment]
OK, separated this into two separate PRs.

File: PhysicsTools/NanoAOD/plugins/GlobalVariablesTableProducer.cc
Comment (arizzi) [inline PhysicsTools/NanoAOD/plugins/GlobalVariablesTableProducer.cc]: @rappoccio can you fix here according to code-checks? i.e. use x.empty()

**Comment by arizzi** [comment]
please also update the DQM as suggested by the bot

**Comment by arizzi** [comment]
@rappoccio so should I merge now and wait for a second PR enabling the mass once DF are merged?

**Comment by rappoccio** [comment]
Let's merge now and then I will make a new PR when the DF change is ready.

**Comment by arizzi** [comment]
added cms-sw#21944 for 10.0.X (as master moved to 10.1.X)

**Pull request merged**
